framework/auth/views.py

```diff
@@ -202,87 +202,158 @@ def forgot_password_post(auth, **kwargs):
 
-        GET '/login/' without any query parameter:
-            redirect to CAS login page with dashboard as target service
+    :param auth: the auth context
+    :param login: `True` if `GET /login`, `False` if `GET /register`
+    :param campaign: a target campaign defined in `auth.campaigns`
+    :param next_url: the service url for CAS login or redirect url for OSF
+    :param logout: used only for `claim_user_registered`
+    :return: data object that contains actions for `auth_register` and `auth_login`
+    :raises: http.BAD_REQUEST
+    """
 
-        GET '/login/?logout=true
-            log user out and redirect to CAS login page with redirect_url or next_url as target service
+    # set target page to `dashboard` if no service url is specified
+    if not next_url:
+        next_url = web_url_for('dashboard', _absolute=True)
 
-        GET '/login/?campaign=instituion:
-            if user is logged in, redirect to 'dashboard'
-            show institution login
+    # Only allow redirects which are relative root or full domain. Disallows external redirects.
+    if not validate_next_url(next_url):
+        raise HTTPError(http.BAD_REQUEST)
 
-        GET '/login/?campaign=prereg:
-            if user is logged in, redirect to prereg home page
-            else show sign up page and notify user to sign in, set next to prereg home page
+    data = {
+        'status_code': http.FOUND if login else http.OK,
+        'next_url': next_url,
+        'campaign': None,
+        'must_login_warning': False,
+    }
 
-        GET '/login/?next=next_url:
-            if user is logged in, redirect to next_url
-            else redirect to CAS login page with next_url as target service
+    if campaign:
+        if validate_campaign(campaign):
+            # GET `/register` or '/login` with `campaign=institution`
+            # unlike other campaigns, institution login serves as an alternative for authentication
+            if campaign == 'institution':
+                data['status_code'] = http.FOUND
+                data['next_url'] = cas.get_login_url(next_url, campaign='institution')
+            # For all other non-institution campaigns
+            else:
+                # `GET /login?campaign=...`
+                if login:
+                    data['next_url'] = web_url_for('auth_register', campaign=campaign)
+                # `GET /register?campaign=...`
+                else:
+                    data['campaign'] = campaign
+                    data['next_url'] = campaigns.campaign_url_for(campaign)
+        else:
+            # invalid campaign
+            raise HTTPError(http.BAD_REQUEST)
+
+    # if user is already logged in, overwrite the actions above and redirect to service url
+    if not logout and auth.logged_in:
+        data['status_code'] = http.FOUND
+
+    # handle `claim_user_registered`
+    # TODO [#OSF-6998]: talk to product about the `must_login_warning`
```

Comment:
update: add `#TODO` ticket
---- 
This is on the fringe of things that actually matter when it comes to changes in login/auth logic, but it might be worth verifying intended behavior of this message with product team. Not blocking